File: chorect-web/src/app/inputLatencyProbe.ts
/**
 * Records how long a touch-down waited between the browser timestamping the event and our
 * handler running (mirror of Android's InputLatencyProbe).
 *
 * Pointer events are dispatched on the main thread, so a slow frame delays the note before
 * the audio graph is involved at all — latency no amount of audio tuning can explain. Keeping
 * it separate says whether a late-feeling tap is an audio problem or a rendering one.
 */
let lastDispatchMs = -1;
let worstDispatchMs = -1;

export function recordInputDispatch(delayMs: number): void {
  const d = Math.max(0, delayMs);
  lastDispatchMs = d;
  if (d > worstDispatchMs) worstDispatchMs = d;
}

export function inputDispatchReport(): { lastMs: number; worstMs: number } {
  return { lastMs: lastDispatchMs, worstMs: worstDispatchMs };
}

File: app/src/main/kotlin/app/guitar/app/InputLatencyProbe.kt
package app.guitar.app

import android.os.SystemClock

/**
 * Records how long a touch-down waited between the finger physically landing and the app's
 * gesture handler running.
 *
 * Touch events are delivered on the UI thread, so a slow frame delays them before the audio
 * engine is even asked for a note — a delay no amount of engine tuning can explain or fix.
 * Isolating it matters: it says whether "the tap sounds late" is an audio problem or a
 * rendering problem.
 *
 * A tiny process-global rather than plumbing a parameter through every FretboardView call
 * site: it is pure diagnostics, written on the UI thread and read by the Settings panel.
 */
object InputLatencyProbe {

    /** Dispatch delay of the most recent touch-down, in ms; -1 before the first tap. */
    @Volatile
    var lastDispatchMs: Long = -1L
        private set

    /** Worst dispatch delay seen this session — jank is intermittent, so the peak is the
     *  number that explains an occasional late-feeling tap. */
    @Volatile
    var worstDispatchMs: Long = -1L
        private set

    /** [eventUptimeMillis] is the pointer event's own timestamp (same clock as
     *  [SystemClock.uptimeMillis]). */
    fun record(eventUptimeMillis: Long) {
        val delay = (SystemClock.uptimeMillis() - eventUptimeMillis).coerceAtLeast(0)
        lastDispatchMs = delay
        if (delay > worstDispatchMs) worstDispatchMs = delay
    }

    fun reset() {
        lastDispatchMs = -1L
        worstDispatchMs = -1L
    }
}
